daml-lf/spec/daml-lf-1.rst

```diff
@@ -593,9 +593,13 @@ Then we can define our kinds, types, and expressions::
 
   Kinds
     k
+      ::= 'nat'                                     -- KindNat
+       | ek                                         -- KindErasable
+
+  Erasable Kind
+    ek
       ::= ⋆                                         -- KindStar
-       |  'nat'                                     -- KindNat
-       |  k₁ → k₂                                   -- KindArrow
+       | k → ek                                     -- KindArrow
```

Comment:
Isn't this is a breaking change because we removed the (uninhabited) kind `k -> nat`? This would be a problem since DAML Engine needs to be able to load DARs which contain type abstractions of this kind.
I would suggest we keep the definition of kinds as is and add the notion of "erasable kinds" as defined above on top of that. Then one can prove that each kind `k` satisfies one of the following three:
1. `k` is erasable.
2. There are `k_1, ..., k_n` with `n >= 1` and `k = k_1 -> k2 -> ... -> k_n -> nat` and hence `k` is uninhabited and can hence be erased at runtime as well.
3. `k = nat`. This is the case where we need to demote from the type to the value level.

Reply:
The problem is that the speedy compiler currently blows up if you try to apply `f` when it has kind `* -> nat`. You can introduce such variables with a forall or type lambda. This makes it an illegal type in some ways but not others. It sounds like you want to do something else in these cases?

For example, the term `/\(f : * -> nat). ADD_NUMERIC @(f Unit)` is perfectly valid under current LF spec, but speedy will balk before it even tries to run it.

Reply:
@sofiafaro-da I understand that the change to the spec you suggested would bring it in line with the implementation. The implementation would not stop accepting DARs it has accepted before. Is that correct? If so, then I'm on board.

Reply:
@sofiafaro-da The implementation will not blow up, because as stated  by @hurryabit the kind `⋆ →nat` is uninhabited. Speedy compiler will handle `/\(f : * -> nat). ADD_NUMERIC @(f Unit)` (producing some garbage code), 
but one cannot run this latter, because there exists no `f` of kind  `⋆ →nat`

I add 3 tests in #6618 (could you merge it with the current PR):
* One checks that `⋆ →nat` is a valid kind
* One checks that `⋆ →nat` is inhabited. 
* `/\(f : * -> nat). ADD_NUMERIC @(f Unit)` can be evaluated without blowing up.

Reply:
We discussed this extensively and came up with a counterexample, where keeping `* -> nat` causes problems in the engine causes a runtime type error:

```
let g: forall(n:nat). Int64 = /\(n: nat). 1
in /\(f: * -> nat). ADD_INT64 10 (g @(f Unit))
```

Because of the interaction between the erasable kind `* -> nat` and non-erasable kind `nat`. We definitely want to avoid this by outlawing `* -> nat` as I proposed here.